✨ Add sessionId getter and sessionStarted callback

fuzzybinary opened the pull request:
### What and why?

sessionStarted does not work on Flutter web, but the session gets updated properly in the UI of the integration test app on re-build of the overlay.

refs: RUMM-3345, #442

### How?

For mobile, the platform channel memoizes the last seen session id and returns it. In cases where the session is sampled out, sessionId returns an empty string.

### Review checklist

- [ ] This pull request has appropriate unit and / or integration tests 
- [ ] This pull request references a Github or JIRA issue

### CI Configuration (optional)
- [x] Run unit tests
- [x] Run integration tests

fuzzybinary committed:
✨ Add sessionId getter and sessionStarted callback

sessionStarted does not work on Flutter web, but the session gets updated properly in the UI of the integration test app on re-build of the overlay.

refs: RUMM-3345, #442

fuzzybinary committed:
💚 Fix unit tests, add more unit tests

0xnm commented on packages/datadog_flutter_plugin/CHANGELOG.md:
```diff
@@ -2,7 +2,7 @@
 ## Unreleased
 
-
+* Add a `sessionId` getter to DdRum, as well as a `sessionStarted` callback. See [#442]
```
```suggestion
* Add a `sessionId` getter to `DdRum`, as well as a `sessionStarted` callback. See [#442]
```

0xnm commented on packages/datadog_flutter_plugin/android/src/main/kotlin/com/datadoghq/flutter/DatadogRumPlugin.kt:
```diff
@@ -341,6 +347,19 @@ class DatadogRumPlugin(
+                "rumSessionStarted",
+                mapOf(
+                    "sessionId" to sessionId,
+                    "sampled" to isDiscarded,
```
shouldn't it be `!isDiscared` here? because if `isDiscarded = true`, it means that session is sampled out, meaning that `sampled` should be `false`.

fuzzybinary replied:
Yeah, I'm going to revisit the name of this variable. `sampled` here is supposed to mean "sampled out" which means `isDiscarded = true` but considering I actually messed it up twice while unit testing, I should have realized it's a bad name.

0xnm commented on packages/datadog_flutter_plugin/lib/src/rum/ddrum.dart:
```diff
@@ -96,6 +96,28 @@ class DdRum {
+  ///
+  /// Session ids are updated when a new session is started in Datadog, usually
+  /// as a result of a user action. If you are using [stopSession], [sessionId] will
+  /// continue to return the last active session until a new session starts.
```
Shouldn't we return `null` instead of there is no active session? what is the reasoning behind returning the last active session when it was stopped? What Browser SDK is doing in this case?

fuzzybinary committed:
🧹 Change sessionId to sessionInfo

On mobile, sessionInfo will always contain the id of the active session with a bool informing users if the session was sampled out.
On web, sessionInfo will be null if no session is active or if the session has been sampled out.

0xnm commented on packages/datadog_flutter_plugin/ios/Classes/DatadogRumPlugin.swift:
```diff
@@ -375,6 +375,18 @@ public class DatadogRumPlugin: NSObject, FlutterPlugin {
         return usrMember
     }
 
+    func rumSessionStarted(_ sessionId: String, _ sampled: Bool) {
```
maybe it is worth to rename `sampled` here as well to be more descriptive? I would personally think its meaning is that session was sampled in.

0xnm commented on packages/datadog_flutter_plugin/lib/src/rum/ddrum.dart:
```diff
@@ -96,6 +96,28 @@ class DdRum {
+
+  /// Set the method to be called when new sessions are started in Datadog RUM.
+  ///
+  /// If a session is sampled out, [sessionId] will be an empty string.
```
I mean above there is the following statement:

```
If you are using [stopSession], [sessionId] will
  /// continue to return the last active session until a new session starts. Untracked
  /// sessions will have [DdRumSessionInfo.isSampledOut] set to `true`
  ///
  /// On web, sessionId is the currently active session, or `null` if the current session
  /// has been sampled out or if there is no active session.
```

So for the consistency, if session was stopped, shouldn't we return null or empty string instead? Not quite sure why we should return last active session when it is not active anymore.

fuzzybinary replied:
You're not going to like the answer .... which is that I don't have any way to know that the session was stopped currently, unless I assume `stopSession` is always called from Flutter

fuzzybinary replied:
Yeah, maybe we close this and delay it until we can either add the sessionId accessor to both native clients, or at least a callback on session stop as well.

fuzzybinary commented:
Closing for now in favor of changing some implementation details on the native side first.
